src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs

```diff
@@ -198,5 +210,82 @@ private static ThreadState ProcFsStateToThreadState(char c)
             }
         }
 
+        internal static bool TryReadStatFile(int pid, out Interop.procfs.ParsedStat stat)
+        {
+            if (!TryGetProcPid(pid, out Interop.procfs.ProcPid procPid))
+            {
+                stat = default;
+                return false;
+            }
+            return Interop.procfs.TryReadStatFile(procPid, out stat);
+        }
+
+        internal static bool TryReadStatusFile(int pid, out Interop.procfs.ParsedStatus status)
+        {
+            if (!TryGetProcPid(pid, out Interop.procfs.ProcPid procPid))
+            {
+                status = default;
+                return false;
+            }
+            return Interop.procfs.TryReadStatusFile(procPid, out status);
+        }
+
+        internal static bool TryReadStatFile(int pid, int tid, out Interop.procfs.ParsedStat stat)
+        {
+            if (!TryGetProcPid(pid, out Interop.procfs.ProcPid procPid))
+            {
+                stat = default;
+                return false;
+            }
+            return Interop.procfs.TryReadStatFile(procPid, tid, out stat);
+        }
+
+        internal static volatile int _procMatchesPidNamespace;
+
+        internal static bool TryGetProcPid(int pid, out Interop.procfs.ProcPid procPid)
+        {
+            // Use '/proc/self' for the current process.
+            if (pid == Environment.ProcessId)
+            {
+                procPid = Interop.procfs.ProcPid.Self;
+                return true;
+            }
+
+            if (ProcMatchesPidNamespace)
+            {
+                procPid = (Interop.procfs.ProcPid)pid;
+                return true;
+            }
+
+            // We can't map a process namespace pid to a procfs pid.
+            procPid = Interop.procfs.ProcPid.Invalid;
+            return false;
+        }
+
+        internal static bool ProcMatchesPidNamespace
+        {
+            get
+            {
+                // _procMatchesPidNamespace is set to:
+                // - 0: when uninitialized,
+                // - 1: '/proc' and the process pid namespace match,
+                // - 2: when they don't match.
+                if (_procMatchesPidNamespace == 0)
+                {
+                    // '/proc/self' is a symlink to the pid used by '/proc' for the current process.
+                    // We compare it with the pid of the current process to see if the '/proc' and pid namespace match up.
+                    int? procSelfPid = null;
+                    if (Interop.Sys.ReadLink($"{Interop.procfs.RootPath}{Interop.procfs.Self}") is string target &&
+                        int.TryParse(target, out int pid))
+                    {
+                        procSelfPid = pid;
+                    }
+                    Debug.Assert(procSelfPid.HasValue);
```

Comment:
`Debug.Assert(procSelfPid.HasValue)` assumes that the value will be always present, but the condition below does not follow this assumption and in such case returns 1 ('/proc' and the process pid namespace match).

Why is that?

Reply:
It's handling the unexpected case where we were unable to determine the proc fs pid.